graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java

```diff
@@ -0,0 +1,178 @@
+                                                 @ApiParam(name = "sort",
+                                                           value = "The field to sort the result on",
+                                                           allowableValues = "name")
+                                                 @QueryParam("sort") String sort,
+                                                 @ApiParam(name = "direction", value = "The sort direction", allowableValues = "asc,desc")
+                                                 @DefaultValue("asc") @QueryParam("direction") SortOrder order) {
+        return entityGroupService.findPaginated(query, page, perPage, order, sort, null);
+    }
+
+    @GET
+    @Path("/{id}")
+    @ApiOperation(value = "Get a single entity group")
+    @ApiResponses(@ApiResponse(code = 404, message = "No such entity group"))
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_READ)
+    public Response get(@ApiParam(name = "id", required = true) @PathParam("id") String id) {
+        try {
+            return Response.ok().entity(entityGroupService.requireEntityGroup(id)).build();
+        } catch (IllegalArgumentException e) {
+            throw new BadRequestException(getRootCauseMessage(e), e);
+        }
+    }
+
+    @GET
+    @Path("/get_for_entity")
+    @ApiOperation(value = "Get a list of entity groups for an entity")
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_READ)
+    public Response getAllForEntity(@ApiParam(name = "entity_type", required = true) @QueryParam("entity_type") String entityType,
+                                    @ApiParam(name = "entity_id", required = true) @QueryParam("entity_id") String entityId) {
+        return Response.ok().entity(entityGroupService.getAllForEntity(entityType, entityId)).build();
+    }
+
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation("Create a new entity group")
+    @ApiResponses(@ApiResponse(code = 400, message = "An entity group already exists with id or name"))
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
+    @AuditEvent(type = AuditEventTypes.ENTITY_GROUP_CREATE)
+    public Response create(@ApiParam(name = "JSON Body") EntityGroup entityGroup) {
+        try {
+            return Response.ok().entity(entityGroupService.create(entityGroup)).build();
+        } catch (DuplicateKeyException | MongoWriteException e) {
+            throw new BadRequestException(StringUtils.f("Entity group '%s' already exists", entityGroup.name()));
+        }
+    }
+
+    @PUT
+    @Path("/{id}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation("Update an entity group")
+    @ApiResponses({
+            @ApiResponse(code = 400, message = "An entity group already exists with id or name"),
+            @ApiResponse(code = 404, message = "No such entity group")
+    })
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
```

Comment:
Permission checks on resources that deal with a specific instance should use per-instance checks, not only annotations. Compare eg. https://github.com/Graylog2/graylog2-server/blob/e53295c35322939ad25a4d3eb67196604bd5d321/graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java#L191